pennylane/qnn.py

```diff
@@ -0,0 +1,205 @@
+
+        @qml.qnode(dev, interface="tf")
+        def circuit(inputs, weights):
+            qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+            qml.templates.StronglyEntanglingLayers(weights, wires=list(range(n_qubits)))
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        weight_shapes = {"weights": (3, n_qubits, 3)}
+        weight_specs = {"weights": {"initializer": "random_uniform"}}
+
+        qnode = qml.qnn.KerasLayer(circuit, weight_shapes, 2, weight_specs)
+
+    The resulting ``qnode`` can be treated as a Keras layer and can be combined with other layers
+    using the `Sequential <https://www.tensorflow.org/api_docs/python/tf/keras/Sequential>`__ or
+    `Model <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`__ APIs.
+
+    Args:
+        qnode (qml.QNode): the PennyLane QNode to be converted into a Keras layer
+        weight_shapes (dict[str, tuple]): a dictionary mapping from all weights used in the QNode to
+            their corresponding sizes
+        output_dim (int): the dimension of data output from the QNode
+        weight_specs (dict[str, dict]): an optional dictionary for users to provide additional
+            specifications for weights used in the QNode, such as the method of parameter
+            initialization. This specification is provided as a dictionary with keys given by the
+            arguments of the `add_weight()
+            <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight>`__
+            (e.g., ``initializer``) method and values being the corresponding specification.
+        **kwargs: additional keyword arguments passed to the `Layer
+            <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer>`__ base class
+    """
+
+    def __init__(
+        self,
+        qnode: QNode,
+        weight_shapes: dict,
+        output_dim: int,
+        weight_specs: Optional[dict] = None,
+        **kwargs
+    ):
+        if not CORRECT_TF_VERSION:
+            raise ImportError("KerasLayer requires TensorFlow version 2 and above")
+
+        self.sig = qnode.func.sig
+
+        if INPUT_ARG not in self.sig:
+            raise TypeError(
+                "QNode must include an argument with name {} for inputting data".format(INPUT_ARG)
+            )
+        if INPUT_ARG in set(weight_shapes.keys()):
+            raise ValueError(
+                "{} argument should not have its dimension specified in "
+                "weight_shapes".format(INPUT_ARG)
+            )
+        if set(weight_shapes.keys()) | {INPUT_ARG} != set(self.sig.keys()):
+            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+        if qnode.func.var_pos:
+            raise TypeError("Cannot have a variable number of positional arguments")
+        if qnode.func.var_keyword:
+            raise TypeError("Cannot have a variable number of keyword arguments")
```

Comment:
Nice validation!

Nitpick: I would put a linebreak between each if statement to break them up for clarity.